src/mender-inventory-mender-configure

```diff
@@ -29,53 +29,122 @@ if ! [ -f "$CONFIG" ]; then
+        return 0
+    fi
+
+    local ret=0
+    curl \
+        -s \
+        -S \
+        -Lf \
+        -X PUT \
+        -H "Authorization: Bearer ${AUTH_TOKEN}" \
+        -H "Content-Type: application/json" \
+        -d "@${CONFIG}" \
+        "${SERVER}/${CONFIG_ENDPOINT}" || ret=$?
 
-DBUS_REPLY=$(echo "$DBUS_RAW" | sed -ne '/^ *string ".*" *$/ {s/^ *string "\(.*\)" *$/\1/; p}')
 
-AUTH_TOKEN="$(echo "$DBUS_REPLY" | sed -ne '1p')"
-SERVER="$(echo "$DBUS_REPLY" | sed -ne '2p')"
+    if [ "$ret" -ne 0 ]; then
+        return 2
+    fi
+    # Update checksum atomically.
+    echo "$COMPUTED_CHECKSUM" > "$CONFIG_CHECKSUM.tmp"
+    sync "$CONFIG_CHECKSUM.tmp"
+    mv "$CONFIG_CHECKSUM.tmp" "$CONFIG_CHECKSUM"
+    sync "$(dirname "$CONFIG_CHECKSUM")"
+    return 0
+}
 
-if [ -z "$AUTH_TOKEN" ]; then
-    echo "mender-inventory-mender-configure: An authentication token could not be obtained over DBus."
-    # Don't produce an error. Most likely mender-auth is not ready or
-    # the device hasn't been accepted
+ret=0
+report_config || ret=$?
+
+if [ "$ret" -eq 0 ]; then
     exit 0
 fi
-if [ -z "$SERVER" ]; then
-    echo "mender-inventory-mender-configure: A server address could not be obtained over DBus."
-    exit 1
-fi
 
-# Include the JWT token in the checksum to report again the
-# configuration to the server in case of new authorization after
-# device decommissioning or connection to a different Mender server
-COMPUTED_CHECKSUM="$( (echo "$AUTH_TOKEN"; cat "$CONFIG") | sha256sum -)"
+# here: expired or inexistent token
 
-# If checksum hasn't changed since previous report, exit silently.
-if [ -f "$CONFIG_CHECKSUM" ] && [ "$COMPUTED_CHECKSUM" = "$(cat "$CONFIG_CHECKSUM")" ]; then
+tmp_file=$(mktemp)
+
+(dbus-monitor \
+    --system \
+    "type='signal',interface='io.mender.Authentication1'" \
+    | sed -n '/JwtTokenStateChange/{n;s/.*string "\(.*\)"/\1/p;q}' \
+    > $tmp_file) &
+PID=$!
+
+cleanup() {
+   pkill -P $PID
+   rm $tmp_file
+}
+trap cleanup EXIT SIGQUIT SIGTERM
+
+FETCH_VALUE=$(sudo dbus-send \
+    --system \
+    --print-reply \
+    --dest=io.mender.AuthenticationManager \
+    /io/mender/AuthenticationManager \
+    io.mender.Authentication1.FetchJwtToken)
+
+if [[ $FETCH_VALUE == *"false"* ]]; then
+    echo "mender-inventory-mender-configure: A JwtToken could not be fetched over DBus"
     exit 0
 fi
 
-# Do the report.
-curl \
-    -s \
-    -S \
-    -Lf \
-    -X PUT \
-    -H "Authorization: Bearer ${AUTH_TOKEN}" \
-    -H "Content-Type: application/json" \
-    -d "@${CONFIG}" \
-    "${SERVER}/${CONFIG_ENDPOINT}"
-
-# Update checksum atomically.
-echo "$COMPUTED_CHECKSUM" > "$CONFIG_CHECKSUM.tmp"
-sync "$CONFIG_CHECKSUM.tmp"
-mv "$CONFIG_CHECKSUM.tmp" "$CONFIG_CHECKSUM"
-sync "$(dirname "$CONFIG_CHECKSUM")"
+timeout 5 tail -fn0 $tmp_file | head -n1 > /dev/null || true
+
+ret=0
+# Let's try again, this time with a JwtToken hopefully
+report_config || ret=$?
```

Comment:
I don't fully understand the `|| true` here. If we don't receive any signal, what is the point of trying again?

I would say instead to do `|| ret=1` so that we exit clean.